File: holoken-core/src/main/java/com/holokenmod/backtrack/hybrid/BackTrackThreadFactory.java
package com.holokenmod.backtrack.hybrid;

import com.holokenmod.grid.Grid;
import com.holokenmod.creation.GridSingleCageCreator;

import java.util.List;
import java.util.concurrent.ThreadFactory;

public class BackTrackThreadFactory implements ThreadFactory {
	private final Grid grid;
	private final List<GridSingleCageCreator> cageCreators;
	private final BackTrackSolutionListener solutionListener;
	private final boolean isPreSolved;
	
	public BackTrackThreadFactory(Grid grid, List<GridSingleCageCreator> cageCreators, boolean isPreSolved, BackTrackSolutionListener solutionListener) {
		this.grid = grid;
		this.cageCreators = cageCreators;
		this.isPreSolved = isPreSolved;
		this.solutionListener = solutionListener;
	}
	
	@Override
	public Thread newThread(Runnable r) {
		return new BackTrackThread(r, createGrid(), cageCreators, isPreSolved, solutionListener);
	}
	
	private Grid createGrid() {
		return grid.copyEmpty();
	}
}

File: holoken-core/src/main/kotlin/com/holokenmod/backtrack/hybrid/BackTrackThreadFactory.kt
package com.holokenmod.backtrack.hybrid

import com.holokenmod.creation.cage.GridSingleCageCreator
import com.holokenmod.grid.Grid
import java.util.concurrent.ThreadFactory

class BackTrackThreadFactory(
    private val grid: Grid,
    private val cageCreators: List<GridSingleCageCreator>,
    private val isPreSolved: Boolean,
    private val solutionListener: BackTrackSolutionListener
) : ThreadFactory {
    override fun newThread(r: Runnable): Thread {
        return BackTrackThread(r, createGrid(), cageCreators, isPreSolved, solutionListener)
    }

    private fun createGrid(): Grid {
        return grid.copyEmpty()
    }
}